tests/node/test_wakunode_sharding.nim

```diff
@@ -0,0 +1,1026 @@
+{.used.}
+
+import
+  std/[options, sequtils, tempfiles],
+  testutils/unittests,
+  chronos,
+  chronicles,
+  stew/shims/net as stewNet
+
+import
+  std/[sequtils, tempfiles],
+  stew/byteutils,
+  stew/shims/net as stewNet,
+  testutils/unittests,
+  chronos,
+  libp2p/switch,
+  libp2p/protocols/pubsub/pubsub
+
+import
+  ../../../waku/[
+    waku_core/topics/pubsub_topic,
+    node/waku_node,
+    common/paging,
+    waku_core,
+    waku_store/common,
+    node/peer_manager,
+    waku_filter_v2/client,
+  ],
+  ../waku_relay/utils,
+  ../waku_archive/archive_utils,
+  ../testlib/[assertions, common, wakucore, wakunode, testasync, futures]
+
+import ../../../waku/waku_relay/protocol
+
+const
+  listenIp = ValidIpAddress.init("0.0.0.0")
+  listenPort = Port(0)
+
+suite "Sharding":
+  var
+    server {.threadvar.}: WakuNode
+    client {.threadvar.}: WakuNode
+
+  asyncSetup:
+    let
+      serverKey = generateSecp256k1Key()
+      clientKey = generateSecp256k1Key()
+
+    server = newTestWakuNode(serverKey, listenIp, listenPort)
+    client = newTestWakuNode(clientKey, listenIp, listenPort)
+
+    await allFutures(server.mountRelay(), client.mountRelay())
+    await allFutures(server.start(), client.start())
+
+  asyncTeardown:
+    await allFutures(server.stop(), client.stop())
+
+  suite "Static Sharding Functionality":
+    asyncTest "Shard Subscription and Peer Dialing":
+      # Given a connected server and client subscribed to the same pubsub shard
+      let
+        topic = "/waku/2/rs/0/1"
+        serverHandler = server.subscribeCompletionHandler(topic)
+        clientHandler = client.subscribeCompletionHandler(topic)
+
+      # await sleepAsync(FUTURE_TIMEOUT)
+
+      await client.connectToNodes(@[server.switch.peerInfo.toRemotePeerInfo()])
+
+      # When the client publishes a message in the subscribed topic
+      discard await client.publish(
+        some(topic),
+        WakuMessage(payload: "message1".toBytes(), contentTopic: "contentTopic"),
+      )
+
+      # Then the server receives the message
+      let
+        serverResult1 = await serverHandler.waitForResult(FUTURE_TIMEOUT)
+        clientResult1 = await clientHandler.waitForResult(FUTURE_TIMEOUT)
+
+      assertResultOk(serverResult1)
+      assertResultOk(clientResult1)
+
+      # When the server publishes a message in the subscribed topic
+      serverHandler.reset()
+      clientHandler.reset()
+      discard await server.publish(
+        some(topic),
+        WakuMessage(payload: "message2".toBytes(), contentTopic: "contentTopic"),
+      )
```

Comment:
What are we testing in the second publish?

Reply:
In one, the client is the publisher, in the other, the server is.